File: backend/routers/host.py
"""
主持人 API 路由
"""
from fastapi import APIRouter, Depends, HTTPException, Header
from sqlalchemy.orm import Session
from backend.database import get_db
from backend.services import ActivityService, VoteService, ParticipantService, ExportService
from backend.routers.websocket import manager
from typing import Optional
import json

router = APIRouter(prefix="/api/host", tags=["host"])

def get_host_session(session_id: Optional[str] = Header(None, alias="X-Session-ID")):
    """验证主持人 session"""
    if not session_id:
        raise HTTPException(status_code=401, detail="未授权")
    return session_id

@router.get("/status")
async def get_host_status(
    session_id: str = Depends(get_host_session),
    db: Session = Depends(get_db)
):
    """获取主持人控制面板状态"""
    # 验证 session
    participant = ParticipantService.get_participant_by_session(db, session_id)
    if not participant or participant.role != 'host':
        raise HTTPException(status_code=403, detail="无权限")
    
    activity = ActivityService.get_current_activity(db)
    if not activity:
        raise HTTPException(status_code=404, detail="没有活动")
    
    # 获取签到人数
    participant_count = ParticipantService.get_participants_count(db, activity.id)
    
    # 获取投票列表
    votes = VoteService.get_votes_by_activity(db, activity.id)
    vote_list = []
    for vote in votes:
        options = json.loads(vote.options) if vote.options else None
        vote_list.append({
            'id': vote.id,
            'title': vote.title,
            'type': vote.type,
            'options': options
        })
    
    return {
        'activity_id': activity.id,
        'activity_name': activity.name,
        'activity_status': activity.status,
        'participant_count': participant_count,
        'votes': vote_list
    }

@router.post("/activity/start")
async def start_activity(
    session_id: str = Depends(get_host_session),
    db: Session = Depends(get_db)
):
    """开始活动"""
    participant = ParticipantService.get_participant_by_session(db, session_id)
    if not participant or participant.role != 'host':
        raise HTTPException(status_code=403, detail="无权限")
    
    activity = ActivityService.get_current_activity(db)
    if not activity:
        raise HTTPException(status_code=404, detail="没有活动")
    
    # 更新活动状态
    ActivityService.update_activity_status(db, activity.id, 'active')
    
    # 广播活动开始
    await manager.broadcast({
        "type": "activity_started",
        "data": {"activity_name": activity.name}
    })
    
    return {"message": "活动已开始"}

@router.post("/vote/start")
async def start_vote(
    vote_id: int,
    session_id: str = Depends(get_host_session),
    db: Session = Depends(get_db)
):
    """开始投票"""
    participant = ParticipantService.get_participant_by_session(db, session_id)
    if not participant or participant.role != 'host':
        raise HTTPException(status_code=403, detail="无权限")
    
    vote = VoteService.get_vote(db, vote_id)
    if not vote:
        raise HTTPException(status_code=404, detail="投票不存在")
    
    # 解析选项
    options = json.loads(vote.options) if vote.options else None
    
    # 广播开始投票
    await manager.broadcast({
        "type": "vote_started",
        "data": {
            "vote_id": vote.id,
            "title": vote.title,
            "type": vote.type,
            "options": options
        }
    })
    
    return {"message": "投票已开始"}

@router.post("/vote/end")
async def end_vote(
    vote_id: int,
    session_id: str = Depends(get_host_session),
    db: Session = Depends(get_db)
):
    """结束投票并显示结果"""
    participant = ParticipantService.get_participant_by_session(db, session_id)
    if not participant or participant.role != 'host':
        raise HTTPException(status_code=403, detail="无权限")
    
    # 获取投票结果
    results = VoteService.get_vote_results(db, vote_id)
    
    # 广播投票结果
    await manager.broadcast({
        "type": "vote_ended",
        "data": results
    })
    
    return {"message": "投票已结束", "results": results}

@router.post("/vote/exit")
async def exit_vote(
    session_id: str = Depends(get_host_session),
    db: Session = Depends(get_db)
):
    """退出投票,回到签到页"""
    participant = ParticipantService.get_participant_by_session(db, session_id)
    if not participant or participant.role != 'host':
        raise HTTPException(status_code=403, detail="无权限")
    
    # 广播退出投票
    await manager.broadcast({
        "type": "vote_exited"
    })
    
    return {"message": "已退出投票"}

@router.post("/activity/end")
async def end_activity(
    session_id: str = Depends(get_host_session),
    db: Session = Depends(get_db)
):
    """结束活动,显示统计"""
    participant = ParticipantService.get_participant_by_session(db, session_id)
    if not participant or participant.role != 'host':
        raise HTTPException(status_code=403, detail="无权限")
    
    activity = ActivityService.get_current_activity(db)
    if not activity:
        raise HTTPException(status_code=404, detail="没有活动")
    
    # 获取活动统计
    summary = ActivityService.get_activity_summary(db, activity.id)
    
    # 更新活动状态
    ActivityService.update_activity_status(db, activity.id, 'ended')
    
    # 广播活动结束
    await manager.broadcast({
        "type": "activity_ended",
        "data": summary
    })
    
    return {"message": "活动已结束", "summary": summary}

@router.post("/activity/close")
async def close_activity(
    session_id: str = Depends(get_host_session),
    db: Session = Depends(get_db)
):
    """退出活动,保存数据并重置"""
    participant = ParticipantService.get_participant_by_session(db, session_id)
    if not participant or participant.role != 'host':
        raise HTTPException(status_code=403, detail="无权限")
    
    activity = ActivityService.get_current_activity(db)
    if not activity:
        raise HTTPException(status_code=404, detail="没有活动")
    
    # 导出数据
    records_file, stats_file = ExportService.export_activity_data(db, activity.id)
    
    # 重置活动数据
    ActivityService.reset_activity_data(db, activity.id)
    
    # 广播活动关闭
    await manager.broadcast({
        "type": "activity_closed"
    })
    
    return {
        "message": "活动已关闭,数据已保存",
        "files": {
            "records": records_file,
            "statistics": stats_file
        }
    }

File: frontend/static/js/host.js
/**
 * 主持人控制页面逻辑
 */

const sessionId = localStorage.getItem('session_id');
let currentVoteId = null;

if (!sessionId) {
    window.location.href = '/signin';
}

// 初始化 WebSocket
const wsClient = new WebSocketClient('host');

// 初始化
document.addEventListener('DOMContentLoaded', () => {
    wsClient.connect();
    loadHostStatus();

    // 注册 WebSocket 消息处理
    wsClient.on('participant_signed_in', (data) => {
        document.getElementById('participant-count').textContent = data.count;
    });

    wsClient.on('activity_started', () => {
        showMessage('活动已开始', 'success');
        updateButtons('active');
    });

    wsClient.on('vote_started', (data) => {
        currentVoteId = data.vote_id;
        showMessage('投票已开始: ' + data.title, 'success');
        updateButtons('voting');
    });

    wsClient.on('vote_ended', () => {
        showMessage('投票已结束', 'info');
        updateButtons('result');
    });

    wsClient.on('vote_exited', () => {
        currentVoteId = null;
        showMessage('已退出当前投票', 'info');
        updateButtons('active');
    });

    wsClient.on('activity_ended', () => {
        showMessage('活动统计已生成', 'success');
        updateButtons('summary');
    });

    wsClient.on('activity_closed', () => {
        showMessage('活动已关闭', 'info');
        localStorage.removeItem('session_id');
        setTimeout(() => window.location.href = '/signin', 2000);
    });

    // 控制按钮事件监听
    document.getElementById('start-activity-btn').addEventListener('click', startActivity);
    document.getElementById('end-vote-btn').addEventListener('click', endVote);
    document.getElementById('exit-vote-btn').addEventListener('click', exitVote);
    document.getElementById('end-activity-btn').addEventListener('click', endActivity);
    document.getElementById('close-activity-btn').addEventListener('click', closeActivity);
});

/**
 * 加载主持人状态
 */
async function loadHostStatus() {
    try {
        const data = await apiRequest('/api/host/status');

        document.getElementById('activity-name').textContent = data.activity_name;
        document.getElementById('participant-count').textContent = data.participant_count;

        // 恢复当前投票 ID
        currentVoteId = data.current_vote_id;

        // 根据活动状态初始化按钮显示
        updateButtons(data.activity_status, currentVoteId !== null);

        renderVoteList(data.votes);
    } catch (error) {
        showMessage('加载失败: ' + error.message, 'error');
    }
}

/**
 * 更新按钮显示状态
 */
function updateButtons(status, isVoting = false) {
    const startBtn = document.getElementById('start-activity-btn');
    const endVoteBtn = document.getElementById('end-vote-btn');
    const exitVoteBtn = document.getElementById('exit-vote-btn');
    const endActivityBtn = document.getElementById('end-activity-btn');
    const closeBtn = document.getElementById('close-activity-btn');

    // 默认全部隐藏
    [startBtn, endVoteBtn, exitVoteBtn, endActivityBtn, closeBtn].forEach(btn => {
        if (btn) btn.style.display = 'none';
    });

    console.log('Updating buttons for status:', status, 'isVoting:', isVoting);

    if (status === 'pending') {
        startBtn.style.display = 'block';
    } else if (status === 'active' || status === 'voting' || status === 'result') {
        // 活动进行中
        if (currentVoteId) {
            // 如果有正在进行的投票或结果展示
            if (status === 'voting' || (status === 'active' && !isVoting)) {
                // 这里的逻辑需要根据 backend 同步的状态来
                // 简化: 只要有 currentVoteId, 就显示 End 和 Exit
                endVoteBtn.style.display = 'block';
                exitVoteBtn.style.display = 'block';
            } else if (status === 'result') {
                exitVoteBtn.style.display = 'block';
            } else {
                // 兜底
                endVoteBtn.style.display = 'block';
                exitVoteBtn.style.display = 'block';
            }
        } else {
            // 没有活动投票, 显示结束活动按钮
            endActivityBtn.style.display = 'block';
        }
    } else if (status === 'summary' || status === 'ended') {
        closeBtn.style.display = 'block';
    }
}

/**
 * 渲染投票列表
 */
function renderVoteList(votes) {
    const voteList = document.getElementById('vote-list');

    if (!votes || votes.length === 0) {
        voteList.innerHTML = '<p class="text-muted">暂无投票项目</p>';
        return;
    }

    voteList.innerHTML = votes.map(vote => `
        <div class="vote-select-item" onclick="selectVote(${vote.id})">
            <div class="vote-select-title">${vote.title}</div>
            <div class="vote-select-type">${getVoteTypeText(vote.type)}</div>
        </div>
    `).join('');
}

function getVoteTypeText(type) {
    const typeMap = {
        'single': '单选题',
        'multiple': '多选题',
        'text': '问答题',
        'rating': '评分题'
    };
    return typeMap[type] || type;
}

/**
 * 选择投票
 */
async function selectVote(voteId) {
    if (currentVoteId) {
        showMessage('当前已有投票正在开启,请先退出', 'error');
        return;
    }

    if (!confirm('确定开始这个投票吗?')) return;

    try {
        await apiRequest('/api/host/vote/start?vote_id=' + voteId, {
            method: 'POST'
        });
    } catch (error) {
        showMessage('开始失败: ' + error.message, 'error');
    }
}

// 绑定到 window 以便 onclick 调用
window.selectVote = selectVote;

async function startActivity() {
    try {
        await apiRequest('/api/host/activity/start', { method: 'POST' });
    } catch (error) {
        showMessage('操作失败: ' + error.message, 'error');
    }
}

async function endVote() {
    if (!currentVoteId) return;
    try {
        await apiRequest('/api/host/vote/end?vote_id=' + currentVoteId, { method: 'POST' });
    } catch (error) {
        showMessage('操作失败: ' + error.message, 'error');
    }
}

async function exitVote() {
    try {
        await apiRequest('/api/host/vote/exit', { method: 'POST' });
    } catch (error) {
        showMessage('操作失败: ' + error.message, 'error');
    }
}

async function endActivity() {
    if (!confirm('确定结束活动并显示总结统计吗?')) return;
    try {
        await apiRequest('/api/host/activity/end', { method: 'POST' });
    } catch (error) {
        showMessage('操作失败: ' + error.message, 'error');
    }
}

async function closeActivity() {
    if (!confirm('确定关闭并同步数据吗?')) return;
    try {
        await apiRequest('/api/host/activity/close', { method: 'POST' });
    } catch (error) {
        showMessage('操作失败: ' + error.message, 'error');
    }
}
